File: AI.Core/ChannelsImplimentation/Channels.cs
using System.Collections.Generic;
using System.Data;
using System.Linq;
using System.Runtime.InteropServices;

namespace AI.Core.ChannelsImplimentation
{
    public sealed class Channels : List<Channel>
    {
        public Channels(int capacity)
        {
            //this. = Enumerable.Repeat((Channel)null, capacity).ToList();
            for(int i = 0; i < capacity; i++)
                Add(null);
        }

        /// <summary>
        /// Setting up values for all channels
        /// </summary>
        public void SetValue(ulong value)
        {
            foreach (Channel currentChannel in this)
            {
                currentChannel.Value = value;
            }
        }

        /// <summary>
        /// Index with ulong instead of int
        /// </summary>
        public Channel this[ulong index]
        {
            get
            {
                return this[(int)index];
            }
            set
            {
                this[(int)index] = value;
            }
        }
    }
}

File: NativeCore/Channels.cpp
#include "Channels.h"
#include "stdafx.h"
#include "stdafx.h"

//C# TO C++ CONVERTER TODO TASK: The .NET System namespace is not available from native C++:
//using namespace System::Collections::Generic;
//C# TO C++ CONVERTER TODO TASK: The .NET System namespace is not available from native C++:
//using namespace System::Data;
//C# TO C++ CONVERTER TODO TASK: The .NET System namespace is not available from native C++:
//using namespace System::Linq;
//C# TO C++ CONVERTER TODO TASK: The .NET System namespace is not available from native C++:
//using namespace System::Runtime::InteropServices;
namespace AI
{
	namespace Core
	{
		namespace ChannelsImplimentation
		{

			Channels::Channels(int capacity)
			{
				//this. = Enumerable.Repeat((Channel)null, capacity).ToList();
				for (int i = 0; i < capacity; i++)
				{
					Add(nullptr);
				}
			}

			void Channels::SetValue(unsigned long long value)
			{
				for (auto currentChannel : this)
				{
					currentChannel->Value = value;
				}
			}

			AI::Core::ChannelsImplimentation::Channel &Channels::operator [](unsigned long long index)
			{
				return this->operator[](static_cast<int>(index));
			}

//C# TO C++ CONVERTER TODO TASK: You cannot specify separate 'set' logic for indexers in native C++:
//			void Channels::setdefault(const unsigned long long &index, Channel *value)
//			{
//				this->operator[](safe_cast<int>(index)) = value;
//			}
		}
	}
}
